
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>14. Creating a DNAseI footprint profile &#8212; NWI-BM066A Computation for Biologists</title>
    
  <link rel="stylesheet" href="../_static/css/index.73d71520a4ca3b99cfee5594769eaaae.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/sphinx-book-theme.2d2078699c18a0efb88233928e1cf6ed.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.3da636dd464baa7582d2.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/language_data.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.be0a4a0c39cd630af62a2fcf693f3f06.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="13. Motifs in DNase I peaks" href="dnase_motifs.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  
  <h1 class="site-logo" id="site-title">NWI-BM066A Computation for Biologists</h1>
  
</a>
</div>

<form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>

<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
  <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="01_introduction.html">
   Introduction
  </a>
 </li>
</ul>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="first_exercises.html">
   1. First exercises
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="lists.html">
   2. Lists
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="control_flow.html">
   3. Control flow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="loops.html">
   4. Loops
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="functions_modules.html">
   5. Functions and modules
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dictionaries.html">
   6. Dictionaries
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="extra_exercises.html">
   7. Assembling the pieces
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="files.html">
   8. Files
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="scripts.html">
   9. Scripts
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="debugging.html">
   10. Debugging
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="pandas.html">
   11. Introduction to pandas DataFrames
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="visualization.html">
   12. Visualization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="dnase_motifs.html">
   13. Motifs in DNase I peaks
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   14. Creating a DNAseI footprint profile
  </a>
 </li>
</ul>

</nav>

 <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="row topbar fixed-top container-xl">
    <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show">
    </div>
    <div class="col pl-2 topbar-main">
        
        <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
            data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
            aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
            title="Toggle navigation" data-toggle="tooltip" data-placement="left">
            <i class="fas fa-bars"></i>
            <i class="fas fa-arrow-left"></i>
            <i class="fas fa-arrow-up"></i>
        </button>
        
        <div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    
    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/content/footprints.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
    
</div>
        <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/vanheeringen-lab/cfb_book"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/vanheeringen-lab/cfb_book/issues/new?title=Issue%20on%20page%20%2Fcontent/footprints.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/vanheeringen-lab/cfb_book/edit/master/content/footprints.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


        <!-- Full screen (wrap in <a> to have style consistency -->
        <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                title="Fullscreen mode"><i
                    class="fas fa-expand"></i></button></a>

        <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/vanheeringen-lab/cfb_book/master?urlpath=tree/content/footprints.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        <a class="colab-button" href="https://colab.research.google.com/github/vanheeringen-lab/cfb_book/blob/master/content/footprints.ipynb"><button type="button" class="btn btn-secondary topbarbtn"
                title="Launch Colab" data-toggle="tooltip" data-placement="left"><img class="colab-button-logo"
                    src="../_static/images/logo_colab.png"
                    alt="Interact on Colab">Colab</button></a>
        
        
    </div>
</div>

    </div>

    <!-- Table of contents -->
    <div class="d-none d-md-block col-md-2 bd-toc show">
        
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i> Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#description">
   14.1. Description
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#data">
     14.1.1. Data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#reading-bam-files">
     14.1.2. Reading BAM files
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#creating-the-profile">
     14.1.3. Creating the profile
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#plotting">
     14.1.4. Plotting
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#approach">
     14.1.5. Approach
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#tricky-issues">
       14.1.5.1. Tricky issues:
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#testing">
     14.1.6. Testing
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#final-exercise">
     14.1.7. Final exercise
    </a>
   </li>
  </ul>
 </li>
</ul>

        </nav>
        
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="creating-a-dnasei-footprint-profile">
<h1><span class="section-number">14. </span>Creating a DNAseI footprint profile<a class="headerlink" href="#creating-a-dnasei-footprint-profile" title="Permalink to this headline">¶</a></h1>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This exercise is <em>optional</em>!</p>
</div>
<div class="section" id="description">
<h2><span class="section-number">14.1. </span>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>The exercise for this week is to write a <strong>script</strong> that plots the cumulative footprint signal in motif sequences.
The script is given two input filenames as arguments on the command line:</p>
<ul class="simple">
<li><p>motif locations in BED format;</p></li>
<li><p>DNAseI sequencing data in indexed BAM format.</p></li>
</ul>
<p>The script needs to find all reads overlapping the motif region,
extended with a certain length,
and take the start position of the reads.
It will plot all read start positions, the DNAseI cutting sites,
averaged over all motif locations.
The result should look something like this:</p>
<p><img alt="An example of a CTCF footprint profile." src="../_images/CTCF.Fetal_Brain.png" /></p>
<div class="section" id="data">
<h3><span class="section-number">14.1.1. </span>Data<a class="headerlink" href="#data" title="Permalink to this headline">¶</a></h3>
<p>The motif matches for three transcription factors in DNAseI hotspots are located in <code class="docutils literal notranslate"><span class="pre">/scratch/cfb/footprints/</span></code>:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/scratch/cfb/footprints/CTCF_motifs.bed</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/scratch/cfb/footprints/NRF1_motifs.bed</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/scratch/cfb/footprints/REST_motifs.bed</span></code></p></li>
</ul>
<p>We have DNAseI sequencing data in BAM format for two different tissues (fetal heart and brain):</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">/scratch/cfb/regmod/dnase/bam/UW.Fetal_Brain.ChromatinAccessibility.merged.bam</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">/scratch/cfb/regmod/dnase/bam/UW.Fetal_Heart.ChromatinAccessibility.merged.bam</span></code></p></li>
</ul>
</div>
<div class="section" id="reading-bam-files">
<h3><span class="section-number">14.1.2. </span>Reading BAM files<a class="headerlink" href="#reading-bam-files" title="Permalink to this headline">¶</a></h3>
<p>To read BAM files, you will use the Python <code class="docutils literal notranslate"><span class="pre">pysam</span></code> module,
see the documentation <a class="reference external" href="http://pysam.readthedocs.org/">here</a>.
Let’s say you have a <code class="docutils literal notranslate"><span class="pre">BAM</span></code> (indexed!) file called <code class="docutils literal notranslate"><span class="pre">example.bam</span></code>.
Here is how to open it with <code class="docutils literal notranslate"><span class="pre">pysam</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pysam</span>
<span class="n">samfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="s2">&quot;example.bam&quot;</span><span class="p">,</span>  <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>You can now retrieve reads mapping to a specific location with the <code class="docutils literal notranslate"><span class="pre">fetch</span></code> method,
which returns an iterator.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">samfile</span><span class="o">.</span><span class="n">fetch</span><span class="p">(</span><span class="s2">&quot;chr6&quot;</span><span class="p">,</span> <span class="mi">170862000</span><span class="p">,</span> <span class="mi">170862010</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">qname</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_name</span><span class="p">,</span> <span class="n">read</span><span class="o">.</span><span class="n">reference_start</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SOLEXA-1GA-1:7:72:651:1250#0 chr6 170861983
SOLEXA-1GA-1:2:66:175:1257#0 chr6 170861983
SOLEXA-1GA-1:4:11:575:1503#0 chr6 170861983
SOLEXA-1GA-2:2:95:69:974#0 chr6 170861991
SOLEXA-1GA-2_2_FC30DG6::5:100:373:1401 chr6 170861993
SOLEXA-1GA-1:2:3:22:793:1306 chr6 170861994
SOLEXA-1GA-2_2_FC30DG6::3:85:1736:96 chr6 170861997
SOLEXA-1GA-2_2_FC30DG6::6:45:1406:1983 chr6 170862000
SOLEXA-1GA-2_2_FC30DG6::5:28:437:474 chr6 170862001
SOLEXA-1GA-1:4:2:416:1943#0 chr6 170862002
SOLEXA-1GA-1:2:5:2:293:1943 chr6 170862006
SOLEXA-1GA-1:2:2:40:1247:887 chr6 170862007
SOLEXA-1GA-2:5:1:790:886#0 chr6 170862008
</pre></div>
</div>
</div>
</div>
<p>See the attributes of the read, an <code class="docutils literal notranslate"><span class="pre">AlignedSegment</span></code> object, <a class="reference external" href="http://pysam.readthedocs.org/en/latest/api.html#pysam.AlignedSegment">here</a>.</p>
</div>
<div class="section" id="creating-the-profile">
<h3><span class="section-number">14.1.3. </span>Creating the profile<a class="headerlink" href="#creating-the-profile" title="Permalink to this headline">¶</a></h3>
<p>You can use a <code class="docutils literal notranslate"><span class="pre">list</span></code> to store the profile.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">profile</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="mi">10</span>
<span class="n">profile</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">profile</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">2</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">):</span>
    <span class="n">profile</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">3</span>

<span class="nb">print</span><span class="p">(</span><span class="n">profile</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[0, 1, 0, 0, 3, 3, 3, 3, 2, 0]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="plotting">
<h3><span class="section-number">14.1.4. </span>Plotting<a class="headerlink" href="#plotting" title="Permalink to this headline">¶</a></h3>
<p>Here is an example how to make a line plot using <code class="docutils literal notranslate"><span class="pre">pandas</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This line is only necessary in Jupyter notebook, not in a script</span>
<span class="o">%</span><span class="k">matplotlib</span> inline

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">sns</span><span class="o">.</span><span class="n">set_style</span><span class="p">(</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>  <span class="c1"># I like this better </span>

<span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;a&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="s2">&quot;b&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="mi">10</span><span class="p">)})</span>

<span class="c1"># plot using pandas</span>
<span class="n">df</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">title</span><span class="o">=</span><span class="s2">&quot;example&quot;</span><span class="p">)</span>

<span class="c1"># customize plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;x label&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;y label&quot;</span><span class="p">)</span>

<span class="c1"># save figure as png (many other formats also possible)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s2">&quot;myplot.png&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Matplotlib is building the font cache; this may take a moment.
</pre></div>
</div>
<img alt="../_images/footprints_8_1.png" src="../_images/footprints_8_1.png" />
</div>
</div>
</div>
<div class="section" id="approach">
<h3><span class="section-number">14.1.5. </span>Approach<a class="headerlink" href="#approach" title="Permalink to this headline">¶</a></h3>
<p>Keep the following things in mind:</p>
<ul class="simple">
<li><p>Think about the problem first, before starting to program.</p></li>
<li><p>Write out the logic of the problem.</p></li>
<li><p>Start out small.</p></li>
<li><p>Think of test sets and create them.</p></li>
<li><p>The BAM files are big! Testing your script on a complete dataset will take too much time. <strong>Don’t copy the whole file to your own directory!</strong>
Create a small test file with few motifs to develop and test your script. What would be a reasonably number? 1? 2? 10? 100? 1000?</p></li>
<li><p>The score of the motif match (higher means a better match) is represented in the 5th column of the BED file.</p></li>
</ul>
<div class="section" id="tricky-issues">
<h4><span class="section-number">14.1.5.1. </span>Tricky issues:<a class="headerlink" href="#tricky-issues" title="Permalink to this headline">¶</a></h4>
<ol class="simple">
<li><p>Motif orientation. A motif can be located on the forward or reverse strand (column 6 in the BED file). Start with only forward motifs, and once that works, add the reverse motifs.</p></li>
<li><p>Read orientation. The start position of the reverse reads is not represented in the <code class="docutils literal notranslate"><span class="pre">pos</span></code> attribute.
Start with only the reads mapped to the forward strand.</p></li>
</ol>
</div>
</div>
<div class="section" id="testing">
<h3><span class="section-number">14.1.6. </span>Testing<a class="headerlink" href="#testing" title="Permalink to this headline">¶</a></h3>
<p>You can use CTCF and/or NRF1 data to test your script and compare your result to published examples.
See <a class="reference external" href="http://genome.cshlp.org/content/21/3/456/F5.expansion.html">Fig. 5</a> of <a class="reference external" href="https://dx.doi.org/10.1101/gr.112656.110">Boyle et al. 2012</a> for CTCF and <a class="reference external" href="http://www.nature.com/nature/journal/v489/n7414/fig_tab/nature11212_F3.html">Fig. 3</a> of <a class="reference external" href="https://dx.doi.org/10.1038/nature11212">Neph et al. 2012</a> for NRF1.</p>
</div>
<div class="section" id="final-exercise">
<h3><span class="section-number">14.1.7. </span>Final exercise<a class="headerlink" href="#final-exercise" title="Permalink to this headline">¶</a></h3>
<p>Create the footprint profile of REST (RE1-Silencing Transcription factor), also known as Neuron-Restrictive Silencer Factor (NRSF),
in both heart and brain fetal tissues.
Compare the resuls and check the literature for REST function.
Can you explain the result?</p>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./content"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        </div>
    </div>
    
    
    <div class='prev-next-bottom'>
        
    <a class='left-prev' id="prev-link" href="dnase_motifs.html" title="previous page"><span class="section-number">13. </span>Motifs in DNase I peaks</a>

    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By Simon van Heeringen, Arjen Brinkman, Kees Albers<br/>
        
            &copy; Copyright 2020.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.3da636dd464baa7582d2.js"></script>


    
  </body>
</html>