{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to pandas DataFrames\n",
    "\n",
    "In this tutorial you are going to use the module `pandas`.\n",
    "We will focus on DataFrames, a convenient object type used in pandas.\n",
    "This tutorial has no exercises, but serves as an introduction to pandas.\n",
    "It will illustrate what you can do with pandas DataFrames. \n",
    "If you want more information on pandas, a (quick) tutorial is found here:\n",
    "\n",
    "[http://pandas.pydata.org/pandas-docs/stable/10min.html](http://pandas.pydata.org/pandas-docs/stable/10min.html)\n",
    "\n",
    "## Importing the pandas module\n",
    "\n",
    "We start by importing the pandas module, for which we use the following command:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that we imported the `pandas` module and renamed it to `pd`. This allows us to use all `pandas` functionality using `pd` (instead of `pandas`). This is very common: it saves typing since `pd` is shorter than `pandas`.\n",
    "\n",
    "## Creating a pandas DataFrame from lists\n",
    "\n",
    "Here we will create a pandas DataFrame containing the following columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_exp1 = [\n",
    "    \"chr1\",\n",
    "    \"chr11\", \n",
    "    \"chr11\", \n",
    "    \"chr11\", \"chr13\", \"chr15\", \"chr17\", \"chr17\", \"chr19\", \"chr19\", \"chr2\", \"chr2\", \"chr3\", \"chr4\", \"chr4\", \"chr4\", \"chr5\", \"chr5\", \"chr6\", \"chr6\"]\n",
    "\n",
    "start_exp1 = [43232415, 74459412, 77774406, 131780211, 20806034, 66795152, 7590368, 26903451, 36705003, 53761044, 74880854, 203102822, 14988512, 4543357, 16077241, 20851791, 130970429, 131563056, 111888063, 143889976]\n",
    "\n",
    "end_exp1 = [43233415, 74460412, 77775406, 131781211, 20807034, 66796152, 7591368, 26904451, 36706003, 53762044, 74881854, 203103822, 14989512, 4544357, 16078241, 20852791, 130971429, 131564056, 111889063, 143890976]\n",
    "\n",
    "refseq_exp1 = [\"NM_024097\",\"NM_001098638\",\"NM_003251\",\"NM_016522\",\"NM_001110221\",\"NR_002441\",\"NM_001126112\",\"NM_005165\",\"NM_007145\",\"NM_173856\",\"NM_004263\",\"NM_003352\",\"NR_046253\",\"NR_037888\",\"NM_006017\",\"NM_147181\",\"NM_001164386\",\"NM_001142599\",\"NM_001164283\",\"NR_027113\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case we create an empty DataFrame first, and we fill it up with the individual columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_peaks_exp1 = pd.DataFrame()\n",
    "\n",
    "df_peaks_exp1['chr'] = chr_exp1\n",
    "df_peaks_exp1['start'] = start_exp1\n",
    "df_peaks_exp1['end'] = end_exp1\n",
    "df_peaks_exp1['refseq'] = refseq_exp1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can check what your DataFrame looks like using  `print()`, or simply type `df_peaks_exp1`. The latter only works in a Jupyter notebook, but the advantage is that you get a nicely formatted table!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      chr      start        end        refseq\n",
      "0    chr1   43232415   43233415     NM_024097\n",
      "1   chr11   74459412   74460412  NM_001098638\n",
      "2   chr11   77774406   77775406     NM_003251\n",
      "3   chr11  131780211  131781211     NM_016522\n",
      "4   chr13   20806034   20807034  NM_001110221\n",
      "5   chr15   66795152   66796152     NR_002441\n",
      "6   chr17    7590368    7591368  NM_001126112\n",
      "7   chr17   26903451   26904451     NM_005165\n",
      "8   chr19   36705003   36706003     NM_007145\n",
      "9   chr19   53761044   53762044     NM_173856\n",
      "10   chr2   74880854   74881854     NM_004263\n",
      "11   chr2  203102822  203103822     NM_003352\n",
      "12   chr3   14988512   14989512     NR_046253\n",
      "13   chr4    4543357    4544357     NR_037888\n",
      "14   chr4   16077241   16078241     NM_006017\n",
      "15   chr4   20851791   20852791     NM_147181\n",
      "16   chr5  130970429  130971429  NM_001164386\n",
      "17   chr5  131563056  131564056  NM_001142599\n",
      "18   chr6  111888063  111889063  NM_001164283\n",
      "19   chr6  143889976  143890976     NR_027113\n"
     ]
    }
   ],
   "source": [
    "print(df_peaks_exp1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>refseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>NM_024097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>NM_001098638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>NM_003251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>NM_016522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>NM_001110221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr15</td>\n",
       "      <td>66795152</td>\n",
       "      <td>66796152</td>\n",
       "      <td>NR_002441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>NM_001126112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>26903451</td>\n",
       "      <td>26904451</td>\n",
       "      <td>NM_005165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr19</td>\n",
       "      <td>36705003</td>\n",
       "      <td>36706003</td>\n",
       "      <td>NM_007145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr19</td>\n",
       "      <td>53761044</td>\n",
       "      <td>53762044</td>\n",
       "      <td>NM_173856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr2</td>\n",
       "      <td>74880854</td>\n",
       "      <td>74881854</td>\n",
       "      <td>NM_004263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr2</td>\n",
       "      <td>203102822</td>\n",
       "      <td>203103822</td>\n",
       "      <td>NM_003352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr3</td>\n",
       "      <td>14988512</td>\n",
       "      <td>14989512</td>\n",
       "      <td>NR_046253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr4</td>\n",
       "      <td>4543357</td>\n",
       "      <td>4544357</td>\n",
       "      <td>NR_037888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr4</td>\n",
       "      <td>16077241</td>\n",
       "      <td>16078241</td>\n",
       "      <td>NM_006017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr4</td>\n",
       "      <td>20851791</td>\n",
       "      <td>20852791</td>\n",
       "      <td>NM_147181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr5</td>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>NM_001164386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr5</td>\n",
       "      <td>131563056</td>\n",
       "      <td>131564056</td>\n",
       "      <td>NM_001142599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chr6</td>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>NM_001164283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr6</td>\n",
       "      <td>143889976</td>\n",
       "      <td>143890976</td>\n",
       "      <td>NR_027113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr      start        end        refseq\n",
       "0    chr1   43232415   43233415     NM_024097\n",
       "1   chr11   74459412   74460412  NM_001098638\n",
       "2   chr11   77774406   77775406     NM_003251\n",
       "3   chr11  131780211  131781211     NM_016522\n",
       "4   chr13   20806034   20807034  NM_001110221\n",
       "5   chr15   66795152   66796152     NR_002441\n",
       "6   chr17    7590368    7591368  NM_001126112\n",
       "7   chr17   26903451   26904451     NM_005165\n",
       "8   chr19   36705003   36706003     NM_007145\n",
       "9   chr19   53761044   53762044     NM_173856\n",
       "10   chr2   74880854   74881854     NM_004263\n",
       "11   chr2  203102822  203103822     NM_003352\n",
       "12   chr3   14988512   14989512     NR_046253\n",
       "13   chr4    4543357    4544357     NR_037888\n",
       "14   chr4   16077241   16078241     NM_006017\n",
       "15   chr4   20851791   20852791     NM_147181\n",
       "16   chr5  130970429  130971429  NM_001164386\n",
       "17   chr5  131563056  131564056  NM_001142599\n",
       "18   chr6  111888063  111889063  NM_001164283\n",
       "19   chr6  143889976  143890976     NR_027113"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peaks_exp1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We make a second DataFrame, which corresponds to a second experiment ('exp2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_exp2 = [\"chr6_dbb_hap3\", \"chr6_mann_hap4\", \"chr6_qbl_hap6\", \"chr6_qbl_hap6\", \"chr6_ssto_hap7\", \"chr7\", \"chr9\", \"chrX\", \"chrX\", \"chrX\"]\n",
    "start_exp2 = [4324541, 1932656, 3218006, 4038208, 3038042, 98029927, 90497271, 71496641, 73506544, 152162171]\n",
    "end_exp2 = [4325541, 1933656, 3219006, 4039208, 3039042, 98030927, 90498271, 71497641, 73507544, 152163171]\n",
    "refseq_exp2 = [\"NM_002121\",\"NR_072994\",\"NR_031601\",\"NM_001290043\",\"NM_002441\",\"NM_018842\",\"NM_178828\",\"NM_001007\",\"NR_030258\",\"NM_001184924\"]\n",
    "\n",
    "df_peaks_exp2 = pd.DataFrame()\n",
    "df_peaks_exp2['chr'] = chr_exp2\n",
    "df_peaks_exp2['start'] = start_exp2\n",
    "df_peaks_exp2['end'] = end_exp2\n",
    "df_peaks_exp2['refseq'] = refseq_exp2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at both DataFrames. The `head()` function shows the first five rows of a DataFrame.\n",
    "\n",
    "They have the same column layout (`chr`, `start`, `end`, `refseq`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>refseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>NM_024097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>NM_001098638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>NM_003251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>NM_016522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>NM_001110221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr      start        end        refseq\n",
       "0   chr1   43232415   43233415     NM_024097\n",
       "1  chr11   74459412   74460412  NM_001098638\n",
       "2  chr11   77774406   77775406     NM_003251\n",
       "3  chr11  131780211  131781211     NM_016522\n",
       "4  chr13   20806034   20807034  NM_001110221"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peaks_exp1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>refseq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6_dbb_hap3</td>\n",
       "      <td>4324541</td>\n",
       "      <td>4325541</td>\n",
       "      <td>NM_002121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6_mann_hap4</td>\n",
       "      <td>1932656</td>\n",
       "      <td>1933656</td>\n",
       "      <td>NR_072994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6_qbl_hap6</td>\n",
       "      <td>3218006</td>\n",
       "      <td>3219006</td>\n",
       "      <td>NR_031601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6_qbl_hap6</td>\n",
       "      <td>4038208</td>\n",
       "      <td>4039208</td>\n",
       "      <td>NM_001290043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6_ssto_hap7</td>\n",
       "      <td>3038042</td>\n",
       "      <td>3039042</td>\n",
       "      <td>NM_002441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              chr    start      end        refseq\n",
       "0   chr6_dbb_hap3  4324541  4325541     NM_002121\n",
       "1  chr6_mann_hap4  1932656  1933656     NR_072994\n",
       "2   chr6_qbl_hap6  3218006  3219006     NR_031601\n",
       "3   chr6_qbl_hap6  4038208  4039208  NM_001290043\n",
       "4  chr6_ssto_hap7  3038042  3039042     NM_002441"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peaks_exp2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "They both represent peaks of a ChIP-seq experiment.\n",
    "\n",
    "For both DataFrames we can add an additional column indicating from which experiment the data comes ('exp1' or 'exp2')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_peaks_exp1['exp'] = 'exp1'\n",
    "df_peaks_exp2['exp'] = 'exp2'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take a look at the two DataFrames, and note how we did this:\n",
    "\n",
    "* We used `[ ]` to define a new colum, with the column name between `[ ]`\n",
    "* We only used only one string (`'exp1'`), but this was automatically expanded to the whole column\n",
    "\n",
    "\n",
    "## Creating a pandas DataFrame using a `dictionary`\n",
    "\n",
    "We can also create a pandas DataFrame from a dictionary. The dictionary should contain *column names as keys* and *lists as values* for the columns. To show this in action, we will use the three following lists:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq = [\"NM_001007\",\"NM_001098638\",\"NM_001110221\",\"NM_001126112\",\"NM_001142599\",\"NM_001164283\",\"NM_001164386\",\"NM_001184924\",\"NM_001290043\",\"NM_002121\",\"NM_002441\",\"NM_003251\",\"NM_003352\",\"NM_004263\",\"NM_005165\",\"NM_006017\",\"NM_007145\",\"NM_016522\",\"NM_018842\",\"NM_024097\",\"NM_147181\",\"NM_173856\",\"NM_178828\"]\n",
    "\n",
    "ensembl = [\"ENSG00000198034\",\"ENSG00000166439\",\"ENSG00000121742\",\"ENSG00000141510\",\"ENSG00000072682\",\"ENSG00000056972\",\"ENSG00000158987\",\"ENSG00000198883\",\"ENSG00000225967\",\"ENSG00000237710\",\"ENSG00000235569\",\"ENSG00000151365\",\"ENSG00000116030\",\"ENSG00000135622\",\"ENSG00000109107\",\"ENSG00000007062\",\"ENSG00000167635\",\"ENSG00000182667\",\"ENSG00000006453\",\"ENSG00000164008\",\"ENSG00000185774\",\"ENSG00000196131\",\"ENSG00000177992\"]\n",
    "\n",
    "genename = [\"RPS4X\",\"RNF169\",\"GJB6\",\"TP53\",\"P4HA2\",\"TRAF3IP2\",\"RAPGEF6\",\"PNMA5\",\"TAP2\",\"HLA-DPB1\",\"MSH5\",\"THRSP\",\"SUMO1\",\"SEMA4F\",\"ALDOC\",\"PROM1\",\"ZNF146\",\"NTM\",\"BAIAP2L1\",\"C1orf50\",\"KCNIP4\",\"VN1R2\",\"SPATA31E1\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we use these lists to create a dictionary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_genes = {\n",
    "    'refseq': refseq,\n",
    "    'ensembl': ensembl,\n",
    "    'genename': genename}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the dictionary, we can create the DataFrame.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes = pd.DataFrame(dict_genes)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the resulting DataFrame called `df_genes`. This DataFrame will help us to map RefSeq to ENSEMBL gene identifiers.\n",
    "\n",
    "Next, we create another DataFrame with gene expression values\n",
    "\n",
    "Again, we will use a dictionary for this.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensembl_1 = [\"ENSG00000237710\", \"ENSG00000182667\", \"ENSG00000121742\", \"ENSG00000158987\", \"ENSG00000196131\", \"ENSG00000151365\", \"ENSG00000056972\", \"ENSG00000235569\", \"ENSG00000006453\", \"ENSG00000198883\", \"ENSG00000225967\", \"ENSG00000166439\", \"ENSG00000135622\", \"ENSG00000007062\", \"ENSG00000109107\", \"ENSG00000167635\", \"ENSG00000164008\", \"ENSG00000141510\", \"ENSG00000177992\"]\n",
    "\n",
    "sample_1 = [None, 1.686, 0.063, 4.222, 0.021, 0.026, 8.169, None, 16.159, 0.08, None, 3.972, 0.844, 26.208, 18.218, 91.049, 3.828, 58.697, 0.024]\n",
    "\n",
    "sample_2 = [None, 1.159, 0.011, 4.291, 0.004, 0.013, 0.306, None, 12.671, 0, None, 6.983, 5.352, 14.336, 5.333, 60.28, 5.137, 47.569, 0]\n",
    "\n",
    "sample_3 = [None, 6.865, 0, 4.727, 0.039, 0.023, 10.459, None, 0.225, 0.011, None, 7.007, 3.402, 6.076, 18.445, 28.716, 4.128, 34.299, 0]\n",
    "\n",
    "sample_4 = [None, 15.691, 2.16, 5.284, 0, 0.181, 1.428, None, 0.184, 0.395, None, 11.842, 4.186, 6.01, 131.013, 25.126, 9.933, 28.18, 0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_expr = {'ensembl': ensembl_1, 'sample1': sample_1, 'sample2': sample_2, 'sample3': sample_3, 'sample4': sample_4}\n",
    "df_expr = pd.DataFrame(dict_expr)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging pandas DataFrames\n",
    "\n",
    "Look at the DataFrames using print (or by just typing the DataFrame name followed by Enter)\n",
    "\n",
    "In case we would have very long DataFrames, can use `df.head()` (where `df` is your DataFrame name)\n",
    "\n",
    "\n",
    "## Merging DataFrames on column values\n",
    "\n",
    "We now have 3 types of data in our 3 DataFrames:\n",
    "\n",
    "* ChIP-seq peaks (`df_peaks`)\n",
    "* A RefSeq to ENSEMBL gene identifier mapping (`df_genes`)\n",
    "* Expression values of genes in 4 samples, with ENSEMBL identifiers (`df_expr`)\n",
    "\n",
    "This is a typical example for any (bioinformatical) analysis in which you want to incorporate different data types, often from different sources. You need to merge them if you want to do anything useful. Pandas DataFrames can be of big help here.\n",
    "\n",
    "So let us try to merge these pieces of data\n",
    "\n",
    "## Merging DataFrames: `concat`\n",
    "\n",
    "Using the command `concat` we can concatenate multiple DataFrames.\n",
    "Let us do this with the 2 ChIP-seq peaks DataFrames:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_peaks = pd.concat([df_peaks_exp1, df_peaks_exp2])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check out the result, and note how the `exp` column still allows us to distinghuish the two experiments.\n",
    "\n",
    "Next, we merge the ChIP-seq peaks with the gene identifier mappings, using the `merge` function\n",
    "\n",
    "## Merging DataFrames: `merge`\n",
    "\n",
    "Here we will merge the three different DataFrames (`df_peaks`, `df_genes`, `df_expr`) using the gene identifiers. Note that we have two different gene identifiers (RefSeq and ENSEMBL).\n",
    "\n",
    "First, we merge `df_peaks` with `df_genes`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_peaks = pd.merge(df_genes, df_peaks, on='refseq')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the resulting DataFrame.\n",
    "Our DataFrame now has the ENSEMBL gene identifiers, so we can also merge the `df_expr`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes_peaks_expr = pd.merge(df_genes_peaks, df_expr, on='ensembl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For convenience, we will use a shorter name for our merged DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all  = df_genes_peaks_expr\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tab completion and help\n",
    "\n",
    "Now that we have merged our data, we can do several neat things.\n",
    "\n",
    "Because `df_all` is a pandas DataFrame, it automatically inherits DataFrame methods.\n",
    "\n",
    "You can get an idea if you type:\n",
    "    \n",
    "    df_all.\n",
    "    \n",
    "And then hit the left \\<Tab\\> key.\n",
    "\n",
    "Notice that you get a lot of options....\n",
    "\n",
    "Suppose you want to know how the `sort_values()` method for a DataFrame works. You can type:\n",
    "\n",
    "    df_all.sort_values?\n",
    "    \n",
    "This HELP menu gives you information about the function. type 'q' or \\<ESC\\> to leave the HELP\n",
    "\n",
    "## Sorting a DataFrame\n",
    "\n",
    "Let us try it out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>genename</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_024097</td>\n",
       "      <td>ENSG00000164008</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.828</td>\n",
       "      <td>5.137</td>\n",
       "      <td>4.128</td>\n",
       "      <td>9.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001098638</td>\n",
       "      <td>ENSG00000166439</td>\n",
       "      <td>RNF169</td>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.972</td>\n",
       "      <td>6.983</td>\n",
       "      <td>7.007</td>\n",
       "      <td>11.842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_016522</td>\n",
       "      <td>ENSG00000182667</td>\n",
       "      <td>NTM</td>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>exp1</td>\n",
       "      <td>1.686</td>\n",
       "      <td>1.159</td>\n",
       "      <td>6.865</td>\n",
       "      <td>15.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_003251</td>\n",
       "      <td>ENSG00000151365</td>\n",
       "      <td>THRSP</td>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001110221</td>\n",
       "      <td>ENSG00000121742</td>\n",
       "      <td>GJB6</td>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_005165</td>\n",
       "      <td>ENSG00000109107</td>\n",
       "      <td>ALDOC</td>\n",
       "      <td>chr17</td>\n",
       "      <td>26903451</td>\n",
       "      <td>26904451</td>\n",
       "      <td>exp1</td>\n",
       "      <td>18.218</td>\n",
       "      <td>5.333</td>\n",
       "      <td>18.445</td>\n",
       "      <td>131.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>TP53</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>exp1</td>\n",
       "      <td>58.697</td>\n",
       "      <td>47.569</td>\n",
       "      <td>34.299</td>\n",
       "      <td>28.180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_173856</td>\n",
       "      <td>ENSG00000196131</td>\n",
       "      <td>VN1R2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53761044</td>\n",
       "      <td>53762044</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_007145</td>\n",
       "      <td>ENSG00000167635</td>\n",
       "      <td>ZNF146</td>\n",
       "      <td>chr19</td>\n",
       "      <td>36705003</td>\n",
       "      <td>36706003</td>\n",
       "      <td>exp1</td>\n",
       "      <td>91.049</td>\n",
       "      <td>60.280</td>\n",
       "      <td>28.716</td>\n",
       "      <td>25.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_004263</td>\n",
       "      <td>ENSG00000135622</td>\n",
       "      <td>SEMA4F</td>\n",
       "      <td>chr2</td>\n",
       "      <td>74880854</td>\n",
       "      <td>74881854</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.844</td>\n",
       "      <td>5.352</td>\n",
       "      <td>3.402</td>\n",
       "      <td>4.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_006017</td>\n",
       "      <td>ENSG00000007062</td>\n",
       "      <td>PROM1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>16077241</td>\n",
       "      <td>16078241</td>\n",
       "      <td>exp1</td>\n",
       "      <td>26.208</td>\n",
       "      <td>14.336</td>\n",
       "      <td>6.076</td>\n",
       "      <td>6.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001164386</td>\n",
       "      <td>ENSG00000158987</td>\n",
       "      <td>RAPGEF6</td>\n",
       "      <td>chr5</td>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>exp1</td>\n",
       "      <td>4.222</td>\n",
       "      <td>4.291</td>\n",
       "      <td>4.727</td>\n",
       "      <td>5.284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001164283</td>\n",
       "      <td>ENSG00000056972</td>\n",
       "      <td>TRAF3IP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>exp1</td>\n",
       "      <td>8.169</td>\n",
       "      <td>0.306</td>\n",
       "      <td>10.459</td>\n",
       "      <td>1.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_002121</td>\n",
       "      <td>ENSG00000237710</td>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>chr6_dbb_hap3</td>\n",
       "      <td>4324541</td>\n",
       "      <td>4325541</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_001290043</td>\n",
       "      <td>ENSG00000225967</td>\n",
       "      <td>TAP2</td>\n",
       "      <td>chr6_qbl_hap6</td>\n",
       "      <td>4038208</td>\n",
       "      <td>4039208</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_002441</td>\n",
       "      <td>ENSG00000235569</td>\n",
       "      <td>MSH5</td>\n",
       "      <td>chr6_ssto_hap7</td>\n",
       "      <td>3038042</td>\n",
       "      <td>3039042</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_018842</td>\n",
       "      <td>ENSG00000006453</td>\n",
       "      <td>BAIAP2L1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>98029927</td>\n",
       "      <td>98030927</td>\n",
       "      <td>exp2</td>\n",
       "      <td>16.159</td>\n",
       "      <td>12.671</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_178828</td>\n",
       "      <td>ENSG00000177992</td>\n",
       "      <td>SPATA31E1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>90497271</td>\n",
       "      <td>90498271</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001184924</td>\n",
       "      <td>ENSG00000198883</td>\n",
       "      <td>PNMA5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152162171</td>\n",
       "      <td>152163171</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          refseq          ensembl   genename             chr      start  \\\n",
       "16     NM_024097  ENSG00000164008    C1orf50            chr1   43232415   \n",
       "0   NM_001098638  ENSG00000166439     RNF169           chr11   74459412   \n",
       "14     NM_016522  ENSG00000182667        NTM           chr11  131780211   \n",
       "9      NM_003251  ENSG00000151365      THRSP           chr11   77774406   \n",
       "1   NM_001110221  ENSG00000121742       GJB6           chr13   20806034   \n",
       "11     NM_005165  ENSG00000109107      ALDOC           chr17   26903451   \n",
       "2   NM_001126112  ENSG00000141510       TP53           chr17    7590368   \n",
       "17     NM_173856  ENSG00000196131      VN1R2           chr19   53761044   \n",
       "13     NM_007145  ENSG00000167635     ZNF146           chr19   36705003   \n",
       "10     NM_004263  ENSG00000135622     SEMA4F            chr2   74880854   \n",
       "12     NM_006017  ENSG00000007062      PROM1            chr4   16077241   \n",
       "4   NM_001164386  ENSG00000158987    RAPGEF6            chr5  130970429   \n",
       "3   NM_001164283  ENSG00000056972   TRAF3IP2            chr6  111888063   \n",
       "7      NM_002121  ENSG00000237710   HLA-DPB1   chr6_dbb_hap3    4324541   \n",
       "6   NM_001290043  ENSG00000225967       TAP2   chr6_qbl_hap6    4038208   \n",
       "8      NM_002441  ENSG00000235569       MSH5  chr6_ssto_hap7    3038042   \n",
       "15     NM_018842  ENSG00000006453   BAIAP2L1            chr7   98029927   \n",
       "18     NM_178828  ENSG00000177992  SPATA31E1            chr9   90497271   \n",
       "5   NM_001184924  ENSG00000198883      PNMA5            chrX  152162171   \n",
       "\n",
       "          end   exp  sample1  sample2  sample3  sample4  \n",
       "16   43233415  exp1    3.828    5.137    4.128    9.933  \n",
       "0    74460412  exp1    3.972    6.983    7.007   11.842  \n",
       "14  131781211  exp1    1.686    1.159    6.865   15.691  \n",
       "9    77775406  exp1    0.026    0.013    0.023    0.181  \n",
       "1    20807034  exp1    0.063    0.011    0.000    2.160  \n",
       "11   26904451  exp1   18.218    5.333   18.445  131.013  \n",
       "2     7591368  exp1   58.697   47.569   34.299   28.180  \n",
       "17   53762044  exp1    0.021    0.004    0.039    0.000  \n",
       "13   36706003  exp1   91.049   60.280   28.716   25.126  \n",
       "10   74881854  exp1    0.844    5.352    3.402    4.186  \n",
       "12   16078241  exp1   26.208   14.336    6.076    6.010  \n",
       "4   130971429  exp1    4.222    4.291    4.727    5.284  \n",
       "3   111889063  exp1    8.169    0.306   10.459    1.428  \n",
       "7     4325541  exp2      NaN      NaN      NaN      NaN  \n",
       "6     4039208  exp2      NaN      NaN      NaN      NaN  \n",
       "8     3039042  exp2      NaN      NaN      NaN      NaN  \n",
       "15   98030927  exp2   16.159   12.671    0.225    0.184  \n",
       "18   90498271  exp2    0.024    0.000    0.000    0.000  \n",
       "5   152163171  exp2    0.080    0.000    0.011    0.395  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.sort_values(by='chr')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "or"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>genename</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NM_002441</td>\n",
       "      <td>ENSG00000235569</td>\n",
       "      <td>MSH5</td>\n",
       "      <td>chr6_ssto_hap7</td>\n",
       "      <td>3038042</td>\n",
       "      <td>3039042</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_001290043</td>\n",
       "      <td>ENSG00000225967</td>\n",
       "      <td>TAP2</td>\n",
       "      <td>chr6_qbl_hap6</td>\n",
       "      <td>4038208</td>\n",
       "      <td>4039208</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NM_002121</td>\n",
       "      <td>ENSG00000237710</td>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>chr6_dbb_hap3</td>\n",
       "      <td>4324541</td>\n",
       "      <td>4325541</td>\n",
       "      <td>exp2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>TP53</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>exp1</td>\n",
       "      <td>58.697</td>\n",
       "      <td>47.569</td>\n",
       "      <td>34.299</td>\n",
       "      <td>28.180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_006017</td>\n",
       "      <td>ENSG00000007062</td>\n",
       "      <td>PROM1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>16077241</td>\n",
       "      <td>16078241</td>\n",
       "      <td>exp1</td>\n",
       "      <td>26.208</td>\n",
       "      <td>14.336</td>\n",
       "      <td>6.076</td>\n",
       "      <td>6.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001110221</td>\n",
       "      <td>ENSG00000121742</td>\n",
       "      <td>GJB6</td>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_005165</td>\n",
       "      <td>ENSG00000109107</td>\n",
       "      <td>ALDOC</td>\n",
       "      <td>chr17</td>\n",
       "      <td>26903451</td>\n",
       "      <td>26904451</td>\n",
       "      <td>exp1</td>\n",
       "      <td>18.218</td>\n",
       "      <td>5.333</td>\n",
       "      <td>18.445</td>\n",
       "      <td>131.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_007145</td>\n",
       "      <td>ENSG00000167635</td>\n",
       "      <td>ZNF146</td>\n",
       "      <td>chr19</td>\n",
       "      <td>36705003</td>\n",
       "      <td>36706003</td>\n",
       "      <td>exp1</td>\n",
       "      <td>91.049</td>\n",
       "      <td>60.280</td>\n",
       "      <td>28.716</td>\n",
       "      <td>25.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_024097</td>\n",
       "      <td>ENSG00000164008</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.828</td>\n",
       "      <td>5.137</td>\n",
       "      <td>4.128</td>\n",
       "      <td>9.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_173856</td>\n",
       "      <td>ENSG00000196131</td>\n",
       "      <td>VN1R2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53761044</td>\n",
       "      <td>53762044</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001098638</td>\n",
       "      <td>ENSG00000166439</td>\n",
       "      <td>RNF169</td>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.972</td>\n",
       "      <td>6.983</td>\n",
       "      <td>7.007</td>\n",
       "      <td>11.842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_004263</td>\n",
       "      <td>ENSG00000135622</td>\n",
       "      <td>SEMA4F</td>\n",
       "      <td>chr2</td>\n",
       "      <td>74880854</td>\n",
       "      <td>74881854</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.844</td>\n",
       "      <td>5.352</td>\n",
       "      <td>3.402</td>\n",
       "      <td>4.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_003251</td>\n",
       "      <td>ENSG00000151365</td>\n",
       "      <td>THRSP</td>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_178828</td>\n",
       "      <td>ENSG00000177992</td>\n",
       "      <td>SPATA31E1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>90497271</td>\n",
       "      <td>90498271</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_018842</td>\n",
       "      <td>ENSG00000006453</td>\n",
       "      <td>BAIAP2L1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>98029927</td>\n",
       "      <td>98030927</td>\n",
       "      <td>exp2</td>\n",
       "      <td>16.159</td>\n",
       "      <td>12.671</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001164283</td>\n",
       "      <td>ENSG00000056972</td>\n",
       "      <td>TRAF3IP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>exp1</td>\n",
       "      <td>8.169</td>\n",
       "      <td>0.306</td>\n",
       "      <td>10.459</td>\n",
       "      <td>1.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001164386</td>\n",
       "      <td>ENSG00000158987</td>\n",
       "      <td>RAPGEF6</td>\n",
       "      <td>chr5</td>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>exp1</td>\n",
       "      <td>4.222</td>\n",
       "      <td>4.291</td>\n",
       "      <td>4.727</td>\n",
       "      <td>5.284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_016522</td>\n",
       "      <td>ENSG00000182667</td>\n",
       "      <td>NTM</td>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>exp1</td>\n",
       "      <td>1.686</td>\n",
       "      <td>1.159</td>\n",
       "      <td>6.865</td>\n",
       "      <td>15.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001184924</td>\n",
       "      <td>ENSG00000198883</td>\n",
       "      <td>PNMA5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152162171</td>\n",
       "      <td>152163171</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          refseq          ensembl   genename             chr      start  \\\n",
       "8      NM_002441  ENSG00000235569       MSH5  chr6_ssto_hap7    3038042   \n",
       "6   NM_001290043  ENSG00000225967       TAP2   chr6_qbl_hap6    4038208   \n",
       "7      NM_002121  ENSG00000237710   HLA-DPB1   chr6_dbb_hap3    4324541   \n",
       "2   NM_001126112  ENSG00000141510       TP53           chr17    7590368   \n",
       "12     NM_006017  ENSG00000007062      PROM1            chr4   16077241   \n",
       "1   NM_001110221  ENSG00000121742       GJB6           chr13   20806034   \n",
       "11     NM_005165  ENSG00000109107      ALDOC           chr17   26903451   \n",
       "13     NM_007145  ENSG00000167635     ZNF146           chr19   36705003   \n",
       "16     NM_024097  ENSG00000164008    C1orf50            chr1   43232415   \n",
       "17     NM_173856  ENSG00000196131      VN1R2           chr19   53761044   \n",
       "0   NM_001098638  ENSG00000166439     RNF169           chr11   74459412   \n",
       "10     NM_004263  ENSG00000135622     SEMA4F            chr2   74880854   \n",
       "9      NM_003251  ENSG00000151365      THRSP           chr11   77774406   \n",
       "18     NM_178828  ENSG00000177992  SPATA31E1            chr9   90497271   \n",
       "15     NM_018842  ENSG00000006453   BAIAP2L1            chr7   98029927   \n",
       "3   NM_001164283  ENSG00000056972   TRAF3IP2            chr6  111888063   \n",
       "4   NM_001164386  ENSG00000158987    RAPGEF6            chr5  130970429   \n",
       "14     NM_016522  ENSG00000182667        NTM           chr11  131780211   \n",
       "5   NM_001184924  ENSG00000198883      PNMA5            chrX  152162171   \n",
       "\n",
       "          end   exp  sample1  sample2  sample3  sample4  \n",
       "8     3039042  exp2      NaN      NaN      NaN      NaN  \n",
       "6     4039208  exp2      NaN      NaN      NaN      NaN  \n",
       "7     4325541  exp2      NaN      NaN      NaN      NaN  \n",
       "2     7591368  exp1   58.697   47.569   34.299   28.180  \n",
       "12   16078241  exp1   26.208   14.336    6.076    6.010  \n",
       "1    20807034  exp1    0.063    0.011    0.000    2.160  \n",
       "11   26904451  exp1   18.218    5.333   18.445  131.013  \n",
       "13   36706003  exp1   91.049   60.280   28.716   25.126  \n",
       "16   43233415  exp1    3.828    5.137    4.128    9.933  \n",
       "17   53762044  exp1    0.021    0.004    0.039    0.000  \n",
       "0    74460412  exp1    3.972    6.983    7.007   11.842  \n",
       "10   74881854  exp1    0.844    5.352    3.402    4.186  \n",
       "9    77775406  exp1    0.026    0.013    0.023    0.181  \n",
       "18   90498271  exp2    0.024    0.000    0.000    0.000  \n",
       "15   98030927  exp2   16.159   12.671    0.225    0.184  \n",
       "3   111889063  exp1    8.169    0.306   10.459    1.428  \n",
       "4   130971429  exp1    4.222    4.291    4.727    5.284  \n",
       "14  131781211  exp1    1.686    1.159    6.865   15.691  \n",
       "5   152163171  exp2    0.080    0.000    0.011    0.395  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.sort_values(by='start')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that these commands did not replace the original DataFrame `df_all`!\n",
    "\n",
    "\n",
    "## Using `[ ]` to obtain DataFrame columns or rows\n",
    "\n",
    "Using `[ ]` we can obtain specific parts of the DataFrame.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        RNF169\n",
       "1          GJB6\n",
       "2          TP53\n",
       "3      TRAF3IP2\n",
       "4       RAPGEF6\n",
       "5         PNMA5\n",
       "6          TAP2\n",
       "7      HLA-DPB1\n",
       "8          MSH5\n",
       "9         THRSP\n",
       "10       SEMA4F\n",
       "11        ALDOC\n",
       "12        PROM1\n",
       "13       ZNF146\n",
       "14          NTM\n",
       "15     BAIAP2L1\n",
       "16      C1orf50\n",
       "17        VN1R2\n",
       "18    SPATA31E1\n",
       "Name: genename, dtype: object"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all['genename']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genename</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RNF169</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GJB6</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TP53</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRAF3IP2</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RAPGEF6</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PNMA5</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TAP2</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-DPB1</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MSH5</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>THRSP</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SEMA4F</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ALDOC</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>PROM1</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ZNF146</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NTM</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BAIAP2L1</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>C1orf50</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>VN1R2</td>\n",
       "      <td>exp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SPATA31E1</td>\n",
       "      <td>exp2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     genename   exp\n",
       "0      RNF169  exp1\n",
       "1        GJB6  exp1\n",
       "2        TP53  exp1\n",
       "3    TRAF3IP2  exp1\n",
       "4     RAPGEF6  exp1\n",
       "5       PNMA5  exp2\n",
       "6        TAP2  exp2\n",
       "7    HLA-DPB1  exp2\n",
       "8        MSH5  exp2\n",
       "9       THRSP  exp1\n",
       "10     SEMA4F  exp1\n",
       "11      ALDOC  exp1\n",
       "12      PROM1  exp1\n",
       "13     ZNF146  exp1\n",
       "14        NTM  exp1\n",
       "15   BAIAP2L1  exp2\n",
       "16    C1orf50  exp1\n",
       "17      VN1R2  exp1\n",
       "18  SPATA31E1  exp2"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## or, for multiple columns:\n",
    "df_all[['genename', 'exp']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using `iloc` we can do the same using indexes\n",
    "\n",
    "One index by default points to **rows** of a DataFrame\n",
    "\n",
    "To get the 2nd row, type:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all.iloc[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Or, for multiple rows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all.iloc[2:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This works very much like indexes of a `list`\n",
    "\n",
    "If you use two indices, you can obtain specific rows *and* columns.\n",
    "\n",
    "The first index is always *rows*, the second *columns*.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>exp1</td>\n",
       "      <td>58.697</td>\n",
       "      <td>47.569</td>\n",
       "      <td>34.299</td>\n",
       "      <td>28.180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>exp1</td>\n",
       "      <td>8.169</td>\n",
       "      <td>0.306</td>\n",
       "      <td>10.459</td>\n",
       "      <td>1.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>exp1</td>\n",
       "      <td>4.222</td>\n",
       "      <td>4.291</td>\n",
       "      <td>4.727</td>\n",
       "      <td>5.284</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start        end   exp  sample1  sample2  sample3  sample4\n",
       "2    7590368    7591368  exp1   58.697   47.569   34.299   28.180\n",
       "3  111888063  111889063  exp1    8.169    0.306   10.459    1.428\n",
       "4  130970429  130971429  exp1    4.222    4.291    4.727    5.284"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.iloc[2:5, 4:11]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing missing values or `NaN`s.\n",
    "\n",
    "In many cases DataFrames contains `NaN`, missing values, and you often have to remove them.\n",
    "\n",
    "We have several such `NaN`s in `df_all`.\n",
    "\n",
    "We can obtain a `boolean` (True/False) for the column called 'sample1' like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     False\n",
       "1     False\n",
       "2     False\n",
       "3     False\n",
       "4     False\n",
       "5     False\n",
       "6      True\n",
       "7      True\n",
       "8      True\n",
       "9     False\n",
       "10    False\n",
       "11    False\n",
       "12    False\n",
       "13    False\n",
       "14    False\n",
       "15    False\n",
       "16    False\n",
       "17    False\n",
       "18    False\n",
       "Name: sample1, dtype: bool"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all['sample1'].isnull()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By using the `~` symbol we can invert this `boolean` column. The `~` acts as `not` for a pandas Series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      True\n",
       "1      True\n",
       "2      True\n",
       "3      True\n",
       "4      True\n",
       "5      True\n",
       "6     False\n",
       "7     False\n",
       "8     False\n",
       "9      True\n",
       "10     True\n",
       "11     True\n",
       "12     True\n",
       "13     True\n",
       "14     True\n",
       "15     True\n",
       "16     True\n",
       "17     True\n",
       "18     True\n",
       "Name: sample1, dtype: bool"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "~df_all['sample1'].isnull()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can use this strategy to remove all rows that have `NaN` in the column `sample1`:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>genename</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001098638</td>\n",
       "      <td>ENSG00000166439</td>\n",
       "      <td>RNF169</td>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.972</td>\n",
       "      <td>6.983</td>\n",
       "      <td>7.007</td>\n",
       "      <td>11.842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001110221</td>\n",
       "      <td>ENSG00000121742</td>\n",
       "      <td>GJB6</td>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>TP53</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>exp1</td>\n",
       "      <td>58.697</td>\n",
       "      <td>47.569</td>\n",
       "      <td>34.299</td>\n",
       "      <td>28.180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001164283</td>\n",
       "      <td>ENSG00000056972</td>\n",
       "      <td>TRAF3IP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>exp1</td>\n",
       "      <td>8.169</td>\n",
       "      <td>0.306</td>\n",
       "      <td>10.459</td>\n",
       "      <td>1.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001164386</td>\n",
       "      <td>ENSG00000158987</td>\n",
       "      <td>RAPGEF6</td>\n",
       "      <td>chr5</td>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>exp1</td>\n",
       "      <td>4.222</td>\n",
       "      <td>4.291</td>\n",
       "      <td>4.727</td>\n",
       "      <td>5.284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001184924</td>\n",
       "      <td>ENSG00000198883</td>\n",
       "      <td>PNMA5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>152162171</td>\n",
       "      <td>152163171</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_003251</td>\n",
       "      <td>ENSG00000151365</td>\n",
       "      <td>THRSP</td>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_004263</td>\n",
       "      <td>ENSG00000135622</td>\n",
       "      <td>SEMA4F</td>\n",
       "      <td>chr2</td>\n",
       "      <td>74880854</td>\n",
       "      <td>74881854</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.844</td>\n",
       "      <td>5.352</td>\n",
       "      <td>3.402</td>\n",
       "      <td>4.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_005165</td>\n",
       "      <td>ENSG00000109107</td>\n",
       "      <td>ALDOC</td>\n",
       "      <td>chr17</td>\n",
       "      <td>26903451</td>\n",
       "      <td>26904451</td>\n",
       "      <td>exp1</td>\n",
       "      <td>18.218</td>\n",
       "      <td>5.333</td>\n",
       "      <td>18.445</td>\n",
       "      <td>131.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_006017</td>\n",
       "      <td>ENSG00000007062</td>\n",
       "      <td>PROM1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>16077241</td>\n",
       "      <td>16078241</td>\n",
       "      <td>exp1</td>\n",
       "      <td>26.208</td>\n",
       "      <td>14.336</td>\n",
       "      <td>6.076</td>\n",
       "      <td>6.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_007145</td>\n",
       "      <td>ENSG00000167635</td>\n",
       "      <td>ZNF146</td>\n",
       "      <td>chr19</td>\n",
       "      <td>36705003</td>\n",
       "      <td>36706003</td>\n",
       "      <td>exp1</td>\n",
       "      <td>91.049</td>\n",
       "      <td>60.280</td>\n",
       "      <td>28.716</td>\n",
       "      <td>25.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_016522</td>\n",
       "      <td>ENSG00000182667</td>\n",
       "      <td>NTM</td>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>exp1</td>\n",
       "      <td>1.686</td>\n",
       "      <td>1.159</td>\n",
       "      <td>6.865</td>\n",
       "      <td>15.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NM_018842</td>\n",
       "      <td>ENSG00000006453</td>\n",
       "      <td>BAIAP2L1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>98029927</td>\n",
       "      <td>98030927</td>\n",
       "      <td>exp2</td>\n",
       "      <td>16.159</td>\n",
       "      <td>12.671</td>\n",
       "      <td>0.225</td>\n",
       "      <td>0.184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_024097</td>\n",
       "      <td>ENSG00000164008</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.828</td>\n",
       "      <td>5.137</td>\n",
       "      <td>4.128</td>\n",
       "      <td>9.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_173856</td>\n",
       "      <td>ENSG00000196131</td>\n",
       "      <td>VN1R2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53761044</td>\n",
       "      <td>53762044</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NM_178828</td>\n",
       "      <td>ENSG00000177992</td>\n",
       "      <td>SPATA31E1</td>\n",
       "      <td>chr9</td>\n",
       "      <td>90497271</td>\n",
       "      <td>90498271</td>\n",
       "      <td>exp2</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          refseq          ensembl   genename    chr      start        end  \\\n",
       "0   NM_001098638  ENSG00000166439     RNF169  chr11   74459412   74460412   \n",
       "1   NM_001110221  ENSG00000121742       GJB6  chr13   20806034   20807034   \n",
       "2   NM_001126112  ENSG00000141510       TP53  chr17    7590368    7591368   \n",
       "3   NM_001164283  ENSG00000056972   TRAF3IP2   chr6  111888063  111889063   \n",
       "4   NM_001164386  ENSG00000158987    RAPGEF6   chr5  130970429  130971429   \n",
       "5   NM_001184924  ENSG00000198883      PNMA5   chrX  152162171  152163171   \n",
       "9      NM_003251  ENSG00000151365      THRSP  chr11   77774406   77775406   \n",
       "10     NM_004263  ENSG00000135622     SEMA4F   chr2   74880854   74881854   \n",
       "11     NM_005165  ENSG00000109107      ALDOC  chr17   26903451   26904451   \n",
       "12     NM_006017  ENSG00000007062      PROM1   chr4   16077241   16078241   \n",
       "13     NM_007145  ENSG00000167635     ZNF146  chr19   36705003   36706003   \n",
       "14     NM_016522  ENSG00000182667        NTM  chr11  131780211  131781211   \n",
       "15     NM_018842  ENSG00000006453   BAIAP2L1   chr7   98029927   98030927   \n",
       "16     NM_024097  ENSG00000164008    C1orf50   chr1   43232415   43233415   \n",
       "17     NM_173856  ENSG00000196131      VN1R2  chr19   53761044   53762044   \n",
       "18     NM_178828  ENSG00000177992  SPATA31E1   chr9   90497271   90498271   \n",
       "\n",
       "     exp  sample1  sample2  sample3  sample4  \n",
       "0   exp1    3.972    6.983    7.007   11.842  \n",
       "1   exp1    0.063    0.011    0.000    2.160  \n",
       "2   exp1   58.697   47.569   34.299   28.180  \n",
       "3   exp1    8.169    0.306   10.459    1.428  \n",
       "4   exp1    4.222    4.291    4.727    5.284  \n",
       "5   exp2    0.080    0.000    0.011    0.395  \n",
       "9   exp1    0.026    0.013    0.023    0.181  \n",
       "10  exp1    0.844    5.352    3.402    4.186  \n",
       "11  exp1   18.218    5.333   18.445  131.013  \n",
       "12  exp1   26.208   14.336    6.076    6.010  \n",
       "13  exp1   91.049   60.280   28.716   25.126  \n",
       "14  exp1    1.686    1.159    6.865   15.691  \n",
       "15  exp2   16.159   12.671    0.225    0.184  \n",
       "16  exp1    3.828    5.137    4.128    9.933  \n",
       "17  exp1    0.021    0.004    0.039    0.000  \n",
       "18  exp2    0.024    0.000    0.000    0.000  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all[~df_all['sample1'].isnull()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Make sure you understand how this worked!\n",
    "\n",
    "We can do the same for getting all rows corresponding to `exp1` (experiment 1):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refseq</th>\n",
       "      <th>ensembl</th>\n",
       "      <th>genename</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exp</th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001098638</td>\n",
       "      <td>ENSG00000166439</td>\n",
       "      <td>RNF169</td>\n",
       "      <td>chr11</td>\n",
       "      <td>74459412</td>\n",
       "      <td>74460412</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.972</td>\n",
       "      <td>6.983</td>\n",
       "      <td>7.007</td>\n",
       "      <td>11.842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_001110221</td>\n",
       "      <td>ENSG00000121742</td>\n",
       "      <td>GJB6</td>\n",
       "      <td>chr13</td>\n",
       "      <td>20806034</td>\n",
       "      <td>20807034</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.011</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_001126112</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>TP53</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7590368</td>\n",
       "      <td>7591368</td>\n",
       "      <td>exp1</td>\n",
       "      <td>58.697</td>\n",
       "      <td>47.569</td>\n",
       "      <td>34.299</td>\n",
       "      <td>28.180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001164283</td>\n",
       "      <td>ENSG00000056972</td>\n",
       "      <td>TRAF3IP2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>111888063</td>\n",
       "      <td>111889063</td>\n",
       "      <td>exp1</td>\n",
       "      <td>8.169</td>\n",
       "      <td>0.306</td>\n",
       "      <td>10.459</td>\n",
       "      <td>1.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001164386</td>\n",
       "      <td>ENSG00000158987</td>\n",
       "      <td>RAPGEF6</td>\n",
       "      <td>chr5</td>\n",
       "      <td>130970429</td>\n",
       "      <td>130971429</td>\n",
       "      <td>exp1</td>\n",
       "      <td>4.222</td>\n",
       "      <td>4.291</td>\n",
       "      <td>4.727</td>\n",
       "      <td>5.284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NM_003251</td>\n",
       "      <td>ENSG00000151365</td>\n",
       "      <td>THRSP</td>\n",
       "      <td>chr11</td>\n",
       "      <td>77774406</td>\n",
       "      <td>77775406</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.013</td>\n",
       "      <td>0.023</td>\n",
       "      <td>0.181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NM_004263</td>\n",
       "      <td>ENSG00000135622</td>\n",
       "      <td>SEMA4F</td>\n",
       "      <td>chr2</td>\n",
       "      <td>74880854</td>\n",
       "      <td>74881854</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.844</td>\n",
       "      <td>5.352</td>\n",
       "      <td>3.402</td>\n",
       "      <td>4.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_005165</td>\n",
       "      <td>ENSG00000109107</td>\n",
       "      <td>ALDOC</td>\n",
       "      <td>chr17</td>\n",
       "      <td>26903451</td>\n",
       "      <td>26904451</td>\n",
       "      <td>exp1</td>\n",
       "      <td>18.218</td>\n",
       "      <td>5.333</td>\n",
       "      <td>18.445</td>\n",
       "      <td>131.013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NM_006017</td>\n",
       "      <td>ENSG00000007062</td>\n",
       "      <td>PROM1</td>\n",
       "      <td>chr4</td>\n",
       "      <td>16077241</td>\n",
       "      <td>16078241</td>\n",
       "      <td>exp1</td>\n",
       "      <td>26.208</td>\n",
       "      <td>14.336</td>\n",
       "      <td>6.076</td>\n",
       "      <td>6.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NM_007145</td>\n",
       "      <td>ENSG00000167635</td>\n",
       "      <td>ZNF146</td>\n",
       "      <td>chr19</td>\n",
       "      <td>36705003</td>\n",
       "      <td>36706003</td>\n",
       "      <td>exp1</td>\n",
       "      <td>91.049</td>\n",
       "      <td>60.280</td>\n",
       "      <td>28.716</td>\n",
       "      <td>25.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NM_016522</td>\n",
       "      <td>ENSG00000182667</td>\n",
       "      <td>NTM</td>\n",
       "      <td>chr11</td>\n",
       "      <td>131780211</td>\n",
       "      <td>131781211</td>\n",
       "      <td>exp1</td>\n",
       "      <td>1.686</td>\n",
       "      <td>1.159</td>\n",
       "      <td>6.865</td>\n",
       "      <td>15.691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_024097</td>\n",
       "      <td>ENSG00000164008</td>\n",
       "      <td>C1orf50</td>\n",
       "      <td>chr1</td>\n",
       "      <td>43232415</td>\n",
       "      <td>43233415</td>\n",
       "      <td>exp1</td>\n",
       "      <td>3.828</td>\n",
       "      <td>5.137</td>\n",
       "      <td>4.128</td>\n",
       "      <td>9.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NM_173856</td>\n",
       "      <td>ENSG00000196131</td>\n",
       "      <td>VN1R2</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53761044</td>\n",
       "      <td>53762044</td>\n",
       "      <td>exp1</td>\n",
       "      <td>0.021</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          refseq          ensembl  genename    chr      start        end  \\\n",
       "0   NM_001098638  ENSG00000166439    RNF169  chr11   74459412   74460412   \n",
       "1   NM_001110221  ENSG00000121742      GJB6  chr13   20806034   20807034   \n",
       "2   NM_001126112  ENSG00000141510      TP53  chr17    7590368    7591368   \n",
       "3   NM_001164283  ENSG00000056972  TRAF3IP2   chr6  111888063  111889063   \n",
       "4   NM_001164386  ENSG00000158987   RAPGEF6   chr5  130970429  130971429   \n",
       "9      NM_003251  ENSG00000151365     THRSP  chr11   77774406   77775406   \n",
       "10     NM_004263  ENSG00000135622    SEMA4F   chr2   74880854   74881854   \n",
       "11     NM_005165  ENSG00000109107     ALDOC  chr17   26903451   26904451   \n",
       "12     NM_006017  ENSG00000007062     PROM1   chr4   16077241   16078241   \n",
       "13     NM_007145  ENSG00000167635    ZNF146  chr19   36705003   36706003   \n",
       "14     NM_016522  ENSG00000182667       NTM  chr11  131780211  131781211   \n",
       "16     NM_024097  ENSG00000164008   C1orf50   chr1   43232415   43233415   \n",
       "17     NM_173856  ENSG00000196131     VN1R2  chr19   53761044   53762044   \n",
       "\n",
       "     exp  sample1  sample2  sample3  sample4  \n",
       "0   exp1    3.972    6.983    7.007   11.842  \n",
       "1   exp1    0.063    0.011    0.000    2.160  \n",
       "2   exp1   58.697   47.569   34.299   28.180  \n",
       "3   exp1    8.169    0.306   10.459    1.428  \n",
       "4   exp1    4.222    4.291    4.727    5.284  \n",
       "9   exp1    0.026    0.013    0.023    0.181  \n",
       "10  exp1    0.844    5.352    3.402    4.186  \n",
       "11  exp1   18.218    5.333   18.445  131.013  \n",
       "12  exp1   26.208   14.336    6.076    6.010  \n",
       "13  exp1   91.049   60.280   28.716   25.126  \n",
       "14  exp1    1.686    1.159    6.865   15.691  \n",
       "16  exp1    3.828    5.137    4.128    9.933  \n",
       "17  exp1    0.021    0.004    0.039    0.000  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all[df_all['exp'] == 'exp1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic calculations\n",
    "\n",
    "There are a lot of methods that allows you to do basic calculation on DataFrames.\n",
    "\n",
    "As an illustration we will calculate the pairwise correlations between the columns `sample1`, `sample2`, `sample3`, and `sample4`.\n",
    "\n",
    "We take the appropriate columns first\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_1 = df_all[['sample1', 'sample2', 'sample3', 'sample4']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can use the pandas `corr` method to calculate all pairwise correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample1</th>\n",
       "      <th>sample2</th>\n",
       "      <th>sample3</th>\n",
       "      <th>sample4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.978523</td>\n",
       "      <td>0.862957</td>\n",
       "      <td>0.253552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>0.978523</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.853166</td>\n",
       "      <td>0.159854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>0.862957</td>\n",
       "      <td>0.853166</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.501015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>0.253552</td>\n",
       "      <td>0.159854</td>\n",
       "      <td>0.501015</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample1   sample2   sample3   sample4\n",
       "sample1  1.000000  0.978523  0.862957  0.253552\n",
       "sample2  0.978523  1.000000  0.853166  0.159854\n",
       "sample3  0.862957  0.853166  1.000000  0.501015\n",
       "sample4  0.253552  0.159854  0.501015  1.000000"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_1.corr()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
